packages/docusaurus-theme-classic/src/theme/DocSidebar/index.tsx

```diff
@@ -174,9 +173,32 @@ function DocSidebarItemLink({
 }
 
 function DocSidebarItem(props) {
+  const {autoCollapsed, parent} = props;
+
+  const collapseItems = (item) => {
+    const isCollapsed = item.collapsed;
+
+    if (!isCollapsed) {
+      item.collapsed = !item.collapsed;
+    } else {
+      if (autoCollapsed) {
+        for (const s in parent) {
+          if (parent[s].type !== 'link') {
+            parent[s].collapsed = true;
+          } else {
+            parent[s].collapsed = false;
+          }
+        }
+      }
+      item.collapsed = !item.collapsed;
+    }
+  };
+
```

Comment:
This part does not seem ideal to me, as it mutates sidebar items in a mutable way. It looks like it works but is not a good practice.


I'll probably add a recursive Object.freeze() in dev so that we are sure this is never done.

Reply:
I still update the object but now I now use it as a state, I know set the updated object to the state, is this a better solution if not what would be a better way of doing such a thing?

Reply:
Unfortunately, it is very important to not mutate these props. I have suggested 2 solutions here: https://github.com/facebook/docusaurus/pull/3811#issuecomment-733646190


The problem with current implementation is:

- When you open this link, the doc is not highlighted/opened anymore in the sidebar: https://deploy-preview-3811--docusaurus-2.netlify.app/classic/docs/migration/manual/ 
- When navigating using navbar from API -> Docs, the "state" (actually, props that should be immutable) is not reset correctly and the sidebar keeps it previously state, and does not highlight the intro document.

It is really important to implement one of the 2 solutions I suggest, and I won't be able to merge anything else, as we must be sure to follow React best practices and to not have bugs like these.

If you feel you can't finish your PR to achieve this goal, just let me know and I'll handle this myself based on your initial code (this means you'll be credited for your contribution attempt).